{% extends "layout.html" %}
{% block breadcrumbs %}
{% endblock %}
{% block content %}
<div class="container">
    <h2>FAQs:</h2>
</div>

<div class="container">

     <div>
        <h3>1. What is VISPR-online?</h3>
        <p>VISPR-online is a web-based interactive framework for CRISPR screens exploration and visualization.
            CRISPR screening helps systematically exploration of the functions of coding and non-coding elements in a genome.
            We previously developed VISPR, bundled with <a href="https://www.ncbi.nlm.nih.gov/pubmed/26673418" target="_blank">MAGeCK</a>,
            an interactive framework to visualize CRISPR screening results
            in a local computer. However, VISPR only supports output of MAGeCK, and requires manual installation and configuration.
            To overcome these limitations, we developed VISPR-online, a web-based server for the visualization of CRISPR screening.
            <br> The advantages of VISPR-online compared with VISPR:
        <ul>
          <li>Interactive gene essentiality exploration.</li>
          <li>Support more popular CRISPR screening analysis tools, including
              <a href="https://www.ncbi.nlm.nih.gov/pubmed/26673418" target="_blank">MAGeCK</a>,
              <a href="https://www.ncbi.nlm.nih.gov/pubmed/27083490" target="_blank">BAGEL</a>, and
              <a href="https://www.ncbi.nlm.nih.gov/pubmed/30674557" target="_blank">JACKS</a>.</li>
          <li>View gRNAs in their genomic context.</li>
          <li>Provide retrieving and sharing session functions.</li>
          <li>Installation and configuration free.</li>
        </ul>
        </p>
     </div>

     <div>
        <h3>2. How to obtain the input files of VISPR-online?</h3>
        <p>VISPR-online is a downstream tool for <a href="https://www.ncbi.nlm.nih.gov/pubmed/25476604" target="_blank">MAGeCK</a>,
             <a href="https://www.ncbi.nlm.nih.gov/pubmed/27083490" target="_blank">BAGEL</a> and
            <a href="https://www.ncbi.nlm.nih.gov/pubmed/30674557" target="_blank">JACKS</a>. The input of VISPR-online is output of
            these tools. For the usage of these tools, please refer to related papers. Besides, we provide a test dataset for these three tools,
            and the download link is on the main page of VISPR-online.
        </p>
     </div>

     <div>
        <h3>3. Format of sgRNA location file.</h3>
        <p>sgRNA location file is a tab-delimited text file without file header. Five columns are required: chromosome name,
        gRNA start, gRNA stop, gRNA ID and strand. The gRNA IDs should be the same with those in read count file.
        </p>
        <pre>
chr2	11143961	11143981	s_7648	-
chr3	186801168	186801188	s_46884	+
chr20	46053591	46053611	s_50992	+
chr3	46705812	46705832	s_57876	+
chr19	46634669	46634689	s_23309	-
chr7	108514201	108514221	s_42887	+</pre>
    </div>

     <div>
        <h3>4. How to resume session and share data?</h3>
        <p>VISPR-online allows users to save sessions to the server and resume them later with session IDs.
            If the "Save session to server" box is checked (Step 6 in "Upload Files" tab), the session will be saved to the server.
            VISPR-online will return a session ID to the user. Users can resume the session later with the ID ("Load Session" tab).
            They can also send the session ID to others to share the data.
        </p>
    </div>

     <div>
        <h3>5. Is VISPR-online open source?</h3>
        <p>VISPR-online is open source under <a href="http://opensource.org/licenses/MIT" target="_blank">MIT</a> license.
            The source code can be downloaded from <a href="https://github.com/lemoncyb/VISPR-online" target="_blank">GitHub</a>.
            Users can install and use VIPSR-online in a local computer or network.
            The README of the repository describes how to install VISPR-online locally.
        </p>
    </div>

    <div>
        <h3>6. Problem caused by browser cache.</h3>
        <p>Brower usually caches the JavaScript, CSS and other data to improve speed and user experience. When
            CRISP-view is updated, user needs to empty the previous cache in the browser. For Safari, press
            <kbd>Opt</kbd>+<kbd>Cmd</kbd>+<kbd>E</kbd> to empty cache.</p>
    </div>

</div>

  <footer>
    <p class="copyright">Â© Wei Li Lab 2019.</p>
  </footer>
{% endblock %}